mapper/orm/src/main/java/org/hibernate/search/mapper/orm/batchindexing/impl/MassIndexerImpl.java

```diff
@@ -0,0 +1,235 @@
+ */
+package org.hibernate.search.mapper.orm.batchindexing.impl;
+
+import java.lang.invoke.MethodHandles;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import org.hibernate.CacheMode;
+import org.hibernate.engine.spi.SessionFactoryImplementor;
+import org.hibernate.search.mapper.orm.MassIndexer;
+import org.hibernate.search.mapper.orm.batchindexing.spi.MassIndexerWithTenant;
+import org.hibernate.search.mapper.orm.impl.HibernateSearchContextService;
+import org.hibernate.search.mapper.orm.logging.impl.Log;
+import org.hibernate.search.mapper.orm.mapping.spi.HibernateOrmMapping;
+import org.hibernate.search.mapper.pojo.work.spi.PojoWorkPlan;
+import org.hibernate.search.util.impl.common.Executors;
+import org.hibernate.search.util.impl.common.LoggerFactory;
+import org.hibernate.search.util.impl.common.StringHelper;
+
+/**
+ * Prepares and configures a BatchIndexingWorkspace to start rebuilding
+ * the indexes for all entity instances in the database.
+ * The type of these entities is either all indexed entities or a
+ * subset, always including all subtypes.
+ *
+ * @author Sanne Grinovero
+ */
+public class MassIndexerImpl implements MassIndexerWithTenant {
+
+	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
+
+	private final PojoWorkPlan workPlan;
+	private final HibernateOrmMapping mapping;
+	private final SessionFactoryImplementor sessionFactory;
+
+	protected final Set<Class<?>> rootEntities;
+
+	// default settings defined here:
+	private int typesToIndexInParallel = 1;
+	private int documentBuilderThreads = 6; //loading the main entity, also responsible for loading of lazy @IndexedEmbedded collections
+	private int objectLoadingBatchSize = 10;
+	private long objectsLimit = 0; //means no limit at all
+	private CacheMode cacheMode = CacheMode.IGNORE;
+	private boolean optimizeAtEnd = true;
+	private boolean purgeAtStart = true;
+	private boolean optimizeAfterPurge = true;
+	//TODO: restore the MassIndexerProgressMonitor
+	private int idFetchSize = 100; //reasonable default as we only load IDs
+	private String tenantIdentifier;
+	private Integer idLoadingTransactionTimeout;
+
+	protected MassIndexerImpl(SessionFactoryImplementor sessionFactory, Class<?>... entities) {
+		this.sessionFactory = sessionFactory;
+		this.mapping = sessionFactory.getServiceRegistry().getService( HibernateSearchContextService.class ).getMapping();
+		this.workPlan = mapping.createSearchManager( sessionFactory.createEntityManager() ).createWorkPlan();
+		this.rootEntities = toRootEntities( entities );
+	}
+
+	/**
+	 * From the set of classes a new set is built containing all indexed
+	 * subclasses, but removing then all subtypes of indexed entities.
+	 *
+	 * @param selection
+	 *
+	 * @return a new set of entities
+	 */
+	private Set<Class<?>> toRootEntities(Class<?>... selection) {
```

Comment:
You made the method non-static, I don't really know why.

When you call a method in a constructor, and it doesn't need to be overridden, I find it's better to make the method static and pass all the necessary parameters. That way, you're sure it won't use any field that is initialized later in the constructor.